src/sail_lean_backend/pretty_print_lean.ml

```diff
@@ -113,6 +117,10 @@ let rec doc_typ (Typ_aux (t, _) as typ) =
   | Typ_app (Id_aux (Id "atom", _), [A_aux (A_nexp (Nexp_aux (Nexp_var ki, _)), _)]) -> string "Int"
   | Typ_tuple ts -> parens (separate_map (space ^^ string "×" ^^ space) doc_typ ts)
   | Typ_id (Id_aux (Id id, _)) -> string id
+  | Typ_app (Id_aux (Id "range", _), [A_aux (A_nexp low, _); A_aux (A_nexp high, _)]) ->
+      (* The heuristics is that 99% of rages are positive, so we try to
+         translate them to Nat. *)
+      if clearly_negative low then string "Int" else string "Nat"
```

Comment:
The idea was to keep two kinds, @Alasdair implemented the `Nat` kind specifically to avoid having heuricstics. Not sure how to solve that for ranges, though.

I think your heuristic doensn't have a one-sided error? `Nexp_neg` can be a double negation, for example

Reply:
So I grepped all the uses of range in the models and my understanding is that for the overwhelming majority of them they are clearly non-negative, and a few clearly include negative elements, so this is an optimistic assumption.

Reply:
The way to do this would be to take an `env` representing the context when converting the type, then you can call `Type_check.prove env (nc_gteq low (nint 0))`. That avoids any syntactic-restrictions, but Lean might need more explicit conversions to be inserted.

Reply:
The rule for conversions is generally that Int to Nat would have to be made explicit while Nat to Int would be implicit

Reply:
@javra what do you think is the best thing to do here?

Reply:
Have you tried using @Alasdair 's suggestion for implementing `clearly_negative`? Otherwise we could just leave it as it is and add a comment to improve the heuristics later on.

But I'd call it `provably_positive` and only return `Nat` if the lower bound can be proven to be positive, otherwise we'd have to use `Int` in order not to run into problems.